### PR TITLE
Allow pass a default value for Event.getField #8304

### LINKED ISSUE #8304: Event API: Allow default value for `getField`
Proposal:

```java
Object getField(FieldReference ref, Object defaultValue);
```

Returns `defaultValue` if the given field reference is not present in the Event.

### PR DESCRIPTION
**Pull request by liketic**
Resolves #8304

### CODE REVIEW

**Comment by jordansissel** [comment]
For completeness, this also needs to get added to RubyEvent and the JRuby proxy:  logstash-core/src/main/java/org/logstash/ext/JrubyEventExtLibrary.java

**Comment by suyograo** [comment]
Jenkins test this please

**Comment by liketic** [comment]
@jordansissel I added the logic for `RubyEvent`, please review again. 😄

**Comment by jakelandis** [comment]
Jenkins test this please

**Comment by jordansissel** [comment]
```
% ./gradlew build
...
% bin/logstash --interactive pry
2017-09-20T23:34:59.829-07:00 [main] ERROR InvocationMethodFactory :
java.lang.RuntimeException: Unknown coercion target: class org.jruby.RubyObject
        at org.jruby.internal.runtime.methods.InvocationMethodFactory.loadArgumentWithCast(InvocationMethodFactory.java:542)
        at org.jruby.internal.runtime.methods.InvocationMethodFactory.loadArguments(InvocationMethodFactory.java:526)
        at org.jruby.internal.runtime.methods.InvocationMethodFactory.createAnnotatedMethodInvocation(InvocationMethodFactory.java:753)
        at org.jruby.internal.runtime.methods.InvocationMethodFactory.addAnnotatedMethodInvoker(InvocationMethodFactory.java:699)
        at org.jruby.internal.runtime.methods.InvocationMethodFactory.getAnnotatedMethodClass(InvocationMethodFactory.java:276)
        at org.jruby.internal.runtime.methods.InvocationMethodFactory.getAnnotatedMethod(InvocationMethodFactory.java:206)
        at org.jruby.RubyModule.defineAnnotatedMethod(RubyModule.java:1069)
        at org.jruby.anno.TypePopulator$ReflectiveTypePopulator.populate(TypePopulator.java:104)
        at org.jruby.RubyModule.defineAnnotatedMethodsIndividually(RubyModule.java:1060)
        at org.jruby.RubyModule.defineAnnotatedMethods(RubyModule.java:947)
        at org.logstash.ext.JrubyEventExtLibrary.load(JrubyEventExtLibrary.java:51)
        at JrubyEventExtService.basicLoad(JrubyEventExtService.java:8)
        at org.jruby.Ruby.loadExtension(Ruby.java:2962)
        at org.jruby.runtime.load.ClassExtensionLibrary.load(ClassExtensionLibrary.java:162)
        at org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:34)
        at org.jruby.runtime.load.LoadService.tryLoadingLibraryOrScript(LoadService.java:891)
        at org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:531)
        at org.jruby.runtime.load.LoadService.require(LoadService.java:402)
        at org.jruby.RubyKernel.requireCommon(RubyKernel.java:962)
        at org.jruby.RubyKernel.require19(RubyKernel.java:955)
        at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrNBlock.call(JavaMethod.java:384)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:338)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:314)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:73)
        at org.jruby.ir.interpreter.Interpreter.INTERPRET_ROOT(Interpreter.java:112)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:99)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:35)
        at org.jruby.ir.IRTranslator.execute(IRTranslator.java:42)
        at org.jruby.Ruby.runInterpreter(Ruby.java:840)
        at org.jruby.Ruby.loadFile(Ruby.java:2877)
        at org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.load(LibrarySearcher.java:243)
        at org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:34)
        at org.jruby.runtime.load.LoadService.tryLoadingLibraryOrScript(LoadService.java:891)
        at org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:531)
        at org.jruby.runtime.load.LoadService.require(LoadService.java:402)
        at org.jruby.RubyKernel.requireCommon(RubyKernel.java:962)
        at org.jruby.RubyKernel.require19(RubyKernel.java:955)
        at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrNBlock.call(JavaMethod.java:384)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:338)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:314)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:73)
        at org.jruby.ir.interpreter.Interpreter.INTERPRET_ROOT(Interpreter.java:112)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:99)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:35)
        at org.jruby.ir.IRTranslator.execute(IRTranslator.java:42)
        at org.jruby.Ruby.runInterpreter(Ruby.java:840)
        at org.jruby.Ruby.loadFile(Ruby.java:2877)
        at org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.load(LibrarySearcher.java:243)
        at org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:34)
        at org.jruby.runtime.load.LoadService.tryLoadingLibraryOrScript(LoadService.java:891)
        at org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:531)
        at org.jruby.runtime.load.LoadService.require(LoadService.java:402)
        at org.jruby.RubyKernel.requireCommon(RubyKernel.java:962)
        at org.jruby.RubyKernel.require19(RubyKernel.java:955)
        at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrNBlock.call(JavaMethod.java:384)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:338)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:314)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:73)
        at org.jruby.ir.interpreter.Interpreter.INTERPRET_ROOT(Interpreter.java:112)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:99)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:35)
        at org.jruby.ir.IRTranslator.execute(IRTranslator.java:42)
        at org.jruby.Ruby.runInterpreter(Ruby.java:840)
        at org.jruby.Ruby.loadFile(Ruby.java:2877)
        at org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.load(LibrarySearcher.java:243)
        at org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:34)
        at org.jruby.runtime.load.LoadService.tryLoadingLibraryOrScript(LoadService.java:891)
        at org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:531)
        at org.jruby.runtime.load.LoadService.require(LoadService.java:402)
        at org.jruby.RubyKernel.requireCommon(RubyKernel.java:962)
        at org.jruby.RubyKernel.require19(RubyKernel.java:955)
        at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrNBlock.call(JavaMethod.java:384)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:338)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:314)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:73)
        at org.jruby.ir.interpreter.Interpreter.INTERPRET_ROOT(Interpreter.java:112)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:99)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:35)
        at org.jruby.ir.IRTranslator.execute(IRTranslator.java:42)
        at org.jruby.Ruby.runInterpreter(Ruby.java:840)
        at org.jruby.Ruby.loadFile(Ruby.java:2877)
        at org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.load(LibrarySearcher.java:243)
        at org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:34)
        at org.jruby.runtime.load.LoadService.tryLoadingLibraryOrScript(LoadService.java:891)
        at org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:531)
        at org.jruby.runtime.load.LoadService.require(LoadService.java:402)
        at org.jruby.RubyKernel.requireCommon(RubyKernel.java:962)
        at org.jruby.RubyKernel.require19(RubyKernel.java:955)
        at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneBlock.call(JavaMethod.java:583)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:101)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:101)
        at org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)
        at home.jls.projects.logstash.lib.bootstrap.environment.RUBY$script(/home/jls/projects/logstash/lib/bootstrap/environment.rb:70)
        at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
        at org.jruby.ir.Compiler$1.load(Compiler.java:95)
        at org.jruby.Ruby.runScript(Ruby.java:828)
        at org.jruby.Ruby.runNormally(Ruby.java:747)
        at org.jruby.Ruby.runNormally(Ruby.java:765)
        at org.jruby.Ruby.runFromMain(Ruby.java:578)
        at org.jruby.Main.doRunFromMain(Main.java:417)
        at org.jruby.Main.internalRun(Main.java:305)
        at org.jruby.Main.run(Main.java:232)
        at org.jruby.Main.main(Main.java:204)
LoadError: load error: jruby_event_ext -- java.lang.RuntimeException: (LoadError) Unknown coercion target: class org.jruby.RubyObject
  require at org/jruby/RubyKernel.java:955
   <main> at /home/jls/projects/logstash/logstash-core/lib/logstash/event.rb:5
  require at org/jruby/RubyKernel.java:955
   <main> at /home/jls/projects/logstash/logstash-core/lib/logstash/pipeline.rb:1
  require at org/jruby/RubyKernel.java:955
   <main> at /home/jls/projects/logstash/logstash-core/lib/logstash/pipeline.rb:8
  require at org/jruby/RubyKernel.java:955
   <main> at /home/jls/projects/logstash/logstash-core/lib/logstash/agent.rb:1
  require at org/jruby/RubyKernel.java:955
   <main> at /home/jls/projects/logstash/lib/bootstrap/environment.rb:70
```

The tests pass, but trying to test manually is failing. Logstash crashes on startup. If I checkout master again and rerun `./gradlew compileJava`  is working.

**Comment by liketic** [comment]
@jordansissel Thanks for your patient. I'm really new for `JRuby`. I will be more careful next time.

**Comment by liketic** [comment]
@jordansissel Can you review this again? Thanks in advance. 😄

**Comment by jordansissel** [comment]
@liketic yep, will do soon.

File: logstash-core/src/test/java/org/logstash/EventTest.java [hunk 1]
```diff
@@ -193,6 +193,26 @@ public void testGetFieldList() throws Exception {
...
 
+
+    @Test
+    public void testGetFieldDefaultValue() throws Exception {
+        Event e = new Event();
+
+        assertNull(e.getField("foo"));
+        assertEquals(1, e.getField("foo", 1));
+        e.setField("foo", 42L);
+        assertEquals(42L, e.getField("foo"));
```
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/EventTest.java]: add this case also: `assertEquals(42L, e.getField("foo", 1))`

File: logstash-core/src/test/java/org/logstash/EventTest.java [hunk 2]
```diff
@@ -193,6 +193,26 @@ public void testGetFieldList() throws Exception {
...
+
+        Map<String, Object> inner = new HashMap<>(2);
+        inner.put("innerFoo", 42L);
+
+        e.setField("bar", inner);
```
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/EventTest.java]: This is probably not an expected use of the event API (setting a value to a HashMap, I mean).

Can you rewrite this to be:

```
e.setField("[bar][innerFoo]", 42L);
```

**Comment by liketic** [comment]
@jordansissel Thanks!

**Comment by andrewvc** [comment]
@jordansissel are there further changes ? It looks like @liketic has made any requested changes.

**Comment by yaauie** [comment]
I've got a couple nitpicks about finals, but it's not worth another set of round-trips to get it rebased etc., so I'll just take it over from here.

**Comment by yaauie** [comment]
@liketic I've rebased your work and opened a replacement PR over here -> https://github.com/elastic/logstash/pull/9115

Once the build completes successfully, we'll get it merged and backported appropriately.

**Pull request closed without merging**